Mathlib/Geometry/Manifold/Diffeomorph.lean

```diff
@@ -562,4 +521,141 @@ theorem contMDiff_transDiffeomorph_left {f : M → M'} :
 @[deprecated (since := "2024-11-21")]
 alias smooth_transDiffeomorph_left := contMDiff_transDiffeomorph_left
 
+end
+
+section Constructions
+
+section Product
+
+/-- Product of two diffeomorphisms. -/
+def prodCongr (h₁ : M ≃ₘ^n⟮I, I'⟯ M') (h₂ : N ≃ₘ^n⟮J, J'⟯ N') :
+    (M × N) ≃ₘ^n⟮I.prod J, I'.prod J'⟯ M' × N' where
+  contMDiff_toFun := (h₁.contMDiff.comp contMDiff_fst).prod_mk (h₂.contMDiff.comp contMDiff_snd)
+  contMDiff_invFun :=
+    (h₁.symm.contMDiff.comp contMDiff_fst).prod_mk (h₂.symm.contMDiff.comp contMDiff_snd)
+  toEquiv := h₁.toEquiv.prodCongr h₂.toEquiv
+
+@[simp]
+theorem prodCongr_symm (h₁ : M ≃ₘ^n⟮I, I'⟯ M') (h₂ : N ≃ₘ^n⟮J, J'⟯ N') :
+    (h₁.prodCongr h₂).symm = h₁.symm.prodCongr h₂.symm :=
+  rfl
+
+@[simp]
+theorem coe_prodCongr (h₁ : M ≃ₘ^n⟮I, I'⟯ M') (h₂ : N ≃ₘ^n⟮J, J'⟯ N') :
+    ⇑(h₁.prodCongr h₂) = Prod.map h₁ h₂ :=
+  rfl
+
+section
+
+variable (I J J' M N N' n)
+
+/-- `M × N` is diffeomorphic to `N × M`. -/
+def prodComm : (M × N) ≃ₘ^n⟮I.prod J, J.prod I⟯ N × M where
+  contMDiff_toFun := contMDiff_snd.prod_mk contMDiff_fst
+  contMDiff_invFun := contMDiff_snd.prod_mk contMDiff_fst
+  toEquiv := Equiv.prodComm M N
+
+@[simp]
+theorem prodComm_symm : (prodComm I J M N n).symm = prodComm J I N M n :=
+  rfl
+
+@[simp]
+theorem coe_prodComm : ⇑(prodComm I J M N n) = Prod.swap :=
+  rfl
+
+/-- `(M × N) × N'` is diffeomorphic to `M × (N × N')`. -/
+def prodAssoc : ((M × N) × N') ≃ₘ^n⟮(I.prod J).prod J', I.prod (J.prod J')⟯ M × N × N' where
+  contMDiff_toFun :=
+    (contMDiff_fst.comp contMDiff_fst).prod_mk
+      ((contMDiff_snd.comp contMDiff_fst).prod_mk contMDiff_snd)
+  contMDiff_invFun :=
+    (contMDiff_fst.prod_mk (contMDiff_fst.comp contMDiff_snd)).prod_mk
+      (contMDiff_snd.comp contMDiff_snd)
+  toEquiv := Equiv.prodAssoc M N N'
+
+end
+
+end Product
+
+section disjointUnion
+
+variable {M' : Type*} [TopologicalSpace M'] [ChartedSpace H M']
+  {M'' : Type*} [TopologicalSpace M''] [ChartedSpace H M'']
+  {N J : Type*} [TopologicalSpace N] [ChartedSpace H N] {J : ModelWithCorners 𝕜 E' H}
+  {N' : Type*} [TopologicalSpace N'] [ChartedSpace H N']
+
+/-- The sum of two diffeomorphisms -/
+def sumCongr (φ : Diffeomorph I J M N n) (ψ : Diffeomorph I J M' N' n) :
+    Diffeomorph I J (M ⊕ M') (N ⊕ N') n where
+  toEquiv := Equiv.sumCongr φ.toEquiv ψ.toEquiv
+  contMDiff_toFun := ContMDiff.sum_map φ.contMDiff_toFun ψ.contMDiff_toFun
+  contMDiff_invFun := ContMDiff.sum_map φ.contMDiff_invFun ψ.contMDiff_invFun
+
+lemma sumCongr_symm_symm (φ : Diffeomorph I J M N n) (ψ : Diffeomorph I J M' N' n) :
+  sumCongr φ.symm ψ.symm = (sumCongr φ ψ).symm := rfl
```

Comment:
bad indentation in this lemma and the next ones (there should be 4 spaces)

Reply:
My mental model was "but the proof is provided on the same line" - which is not covered by the style guide, though. Thanks for pointing this out, changed.